core/src/test/java/kafka/test/api/ShareConsumerTest.java

```diff
@@ -1806,6 +1816,181 @@ public void testShareAutoOffsetResetByDurationInvalidFormat() throws Exception {
+                ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
+                    tpMulti.topic(),
+                    tpMulti.partition(),
+                    null,
+                    "key".getBytes(),
+                    "value".getBytes()
+                );
+                IntStream.range(0, 10).forEach(__ -> producer.send(record));
+                producer.flush();
+            }
+
+            // consume messages should only be possible if partition and share coord has moved
+            // from shutdown broker since we are only producing to partition 0 of topic.
+            try (ShareConsumer<byte[], byte[]> shareConsumer = createShareConsumer(groupId)) {
+                shareConsumer.subscribe(List.of(topicName));
+                alterShareAutoOffsetReset(groupId, "earliest");
+                ConsumerRecords<byte[], byte[]> records = shareConsumer.poll(Duration.ofMillis(5000));
+                assertEquals(20, records.count());
+            }
+
+            verifyShareGroupStateTopicRecordsProduced();
+        }
+    }
+
+    @ClusterTest(
+        brokers = 3,
+        serverProperties = {
+            @ClusterConfigProperty(key = "auto.create.topics.enable", value = "false"),
+            @ClusterConfigProperty(key = "group.coordinator.rebalance.protocols", value = "classic,consumer,share"),
+            @ClusterConfigProperty(key = "group.share.enable", value = "true"),
+            @ClusterConfigProperty(key = "group.share.partition.max.record.locks", value = "10000"),
+            @ClusterConfigProperty(key = "group.share.record.lock.duration.ms", value = "15000"),
+            @ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "3"),
+            @ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "3"),
+            @ClusterConfigProperty(key = "share.coordinator.state.topic.min.isr", value = "1"),
+            @ClusterConfigProperty(key = "share.coordinator.state.topic.num.partitions", value = "3"),
+            @ClusterConfigProperty(key = "share.coordinator.state.topic.replication.factor", value = "3"),
+            @ClusterConfigProperty(key = "transaction.state.log.min.isr", value = "1"),
+            @ClusterConfigProperty(key = "transaction.state.log.replication.factor", value = "3"),
+            @ClusterConfigProperty(key = "unstable.api.versions.enable", value = "true")
+        }
+    )
+    public void testComplexShareConsumer() throws Exception {
+        setup();
+        String topicName = "multipart";
+        String groupId = "multipartGrp";
+        createTopic(topicName, 3, 3);
+        TopicPartition multiTp = new TopicPartition(topicName, 0);
+
+        ExecutorService executer = Executors.newCachedThreadPool();
+
+        AtomicBoolean prodDone = new AtomicBoolean(false);
+        AtomicInteger sentCount = new AtomicInteger(0);
+
+        // produce messages until we want
+        executer.execute(() -> {
+            while (!prodDone.get()) {
```

Comment:
This makes a separate producer per loop iteration. I suggest re-using the producer across loop iterations. You can still use try-with-resources.